### PR TITLE
[RLlib] Move learning_starts logic into execution plans

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

* `learning_starts` should be renamed to something more descriptive: `num_steps_sampled_before_learning_starts`

* Should be moved out of replay buffer config according to our philosophy: Algorithm should define what should happen when, but NOT how it should happen. `num_steps_sampled_before_learning_starts` answers the "when" and "what" questions and should thus be handled and configured on the top Algo level (not inside replay buffers).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: rllib/BUILD
Comment (ArturNiederfahrenhorst) [inline rllib/BUILD:3056]: This does not conform with the buffer config arg name, but seemed more expressive to me as a standalone cli arg.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (kouroshHakha) [inline rllib/algorithms/alpha_zero/alpha_zero.py:379]: **question**: Are we removing execution plans in this PR? why not create a separate PR for it? Always isolate the PRs.

File: rllib/algorithms/cql/tests/test_cql.py
Comment (kouroshHakha) [inline rllib/algorithms/cql/tests/test_cql.py:61]: CQL does not need replay_buffer anymore.

File: rllib/examples/offline_rl.py
Comment (kouroshHakha) [inline rllib/examples/offline_rl.py:41]: CQL does not need RB.

File: rllib/tuned_examples/cql/halfcheetah-bc.yaml
Comment (kouroshHakha) [inline rllib/tuned_examples/cql/halfcheetah-bc.yaml:31]: CQL does not need RB

**Review by kouroshHakha — CHANGES_REQUESTED** [review]
**TODO** Please remove the replay buffer configs all together from CQL / CRR algorithms. To identify all of them (As you and I may have missed them) one thing you can do is raise an error in CRR / CQL 'Algorithm.setup()' when self.config['replay_buffer'] is not empty and run all the tests with it. 

On a higher level note, I do not think that `min_size` or `learning_start` or `replay_start` or what ever should be a config of the replay_buffer. If we want to conform to our API design principle which is Algorithm controls "When" certain things should happen, and replay buffer controls "What" should happen, we need to put this periodic event in the algorithm and not hide it under the `sample()` method of a replay buffer. Even though logically they are both correct putting it inside replay buffer adds to the obscurity of our API design. User need to kind of know that when there is not enough samples the replay buffer would return empty batches. So my suggestion is to put it in our back-log to move the logic entirely out of replay buffer. With this re-design we can then name it `replay_starts` or something similar

```
SAC(Algorithm):

    def training_step():
            if cur_ts > self.config['replay_starts']:
                     # sample replay buffer and update the network.
```

**Note**: What I suggested is an extremely low priority P2 for now, since we have other API redos in the pipeline

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/alpha_zero/alpha_zero.py:379]: This was actually removed in https://github.com/ray-project/ray/pull/25624, which got reverted but somehow that revert seemed to have ended up in a bad merge by me. Thanks for noticing! :)

File: rllib/examples/offline_rl.py
Comment (ArturNiederfahrenhorst) [reply rllib/examples/offline_rl.py:41]: @kouroshHakha The complete example is actually broken.
See https://github.com/ray-project/ray/issues/26665.
Not fixing this in this PR though.

File: rllib/tuned_examples/cql/halfcheetah-cql.yaml
Comment (ArturNiederfahrenhorst) [reply rllib/tuned_examples/cql/halfcheetah-cql.yaml:33]: See https://github.com/ray-project/ray/issues/26665.

File: rllib/tuned_examples/cql/hopper-bc.yaml
Comment (ArturNiederfahrenhorst) [reply rllib/tuned_examples/cql/hopper-bc.yaml:31]: See https://github.com/ray-project/ray/issues/26665.

File: rllib/tuned_examples/cql/hopper-cql.yaml
Comment (ArturNiederfahrenhorst) [reply rllib/tuned_examples/cql/hopper-cql.yaml:31]: See https://github.com/ray-project/ray/issues/26665.

**Comment by avnishn** [comment]
can someone rename this pr. This thing goes way beyond renaming a parameter IIUC XD

File: rllib/algorithms/ddpg/tests/test_ddpg.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/ddpg/tests/test_ddpg.py:133]: We now have to add this to a couple of tests. Before, the training iteration function would just run through with an empty sample batch if the buffer was not yet "ready".  With the changes to our execution plans, we now would wait for a while for enough samples to be collected with is nonsense for the compilation tests.

File: rllib/algorithms/dreamer/dreamer.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/dreamer/dreamer.py:412]: I moved this up to the sampling steps of this training step, for the sake of keeping this training step clean. Please step forward if you have thoughts about algorithmic correctness - not an expert on dreamer.

File: rllib/examples/offline_rl.py
Comment (ArturNiederfahrenhorst) [reply rllib/examples/offline_rl.py:41]: Adding to this so that it will not block this PR: The example is not part of CI and so was forgotten when we changed CQL.

File: rllib/examples/two_trainer_workflow.py
Comment (ArturNiederfahrenhorst) [inline rllib/examples/two_trainer_workflow.py:115]: This is sort of a manual implementation of the learning starts logic. Since this example implements custom training logic, we put it here in a simplified verison.

File: rllib/tests/test_execution.py
Comment (ArturNiederfahrenhorst) [inline rllib/tests/test_execution.py:211]: test_execution will be gone soon. The logic is gone from the replay buffers and is now covered by our tests covering training_steps.

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
Self-reviewed.

**Review by sven1977 — APPROVED** [review]
Looks good to me now. Thanks for your patience on this @ArturNiederfahrenhorst , and thanks @kouroshHakha for raising the very valid objection and arguments to why we should leave this in the top-level algo config , instead of moving it into the replay buffer config.

File: release/rllib_tests/learning_tests/yaml_files/td3/td3-halfcheetahbulletenv-v0.yaml
Comment (kouroshHakha) [inline release/rllib_tests/learning_tests/yaml_files/td3/td3-halfcheetahbulletenv-v0.yaml:12]: replay_buffer_config needs to be removed entirely then.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (kouroshHakha) [inline rllib/algorithms/alpha_zero/alpha_zero.py:118]: I don't wanna wreck this for you @ArturNiederfahrenhorst 🥲 but is this number of env steps / agent steps?

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (kouroshHakha) [inline rllib/algorithms/alpha_zero/alpha_zero.py:225]: I assume it is number of environment timesteps, can we make the documentation more explicit here?

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (kouroshHakha) [reply rllib/algorithms/alpha_zero/alpha_zero.py:118]: Ok it only needs a bit more explicit documentation.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (kouroshHakha) [reply rllib/algorithms/alpha_zero/alpha_zero.py:225]: see my comment below as well.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (kouroshHakha) [inline rllib/algorithms/alpha_zero/alpha_zero.py:356]: See the problem? I had to read this code to understand that it can actually be both and it really depends on the _agent_step_by parameter which one will be used.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/alpha_zero/alpha_zero.py:118]: It depends on how you count, just like with STEPS_TRAINED_COUNTER and so on. We generally refer to num_steps_*** as whatever user has chosen to count in.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (kouroshHakha) [inline rllib/algorithms/alpha_zero/alpha_zero.py:172]: Should this be part of AlgorithmConfig instead of each specific algorithm? We need to put it in a place that is generally accessible by all other configs.

File: rllib/algorithms/ddpg/tests/test_ddpg.py
Comment (kouroshHakha) [inline rllib/algorithms/ddpg/tests/test_ddpg.py:40]: why do you have to specify this twice? you have already done this in line 37?

File: rllib/algorithms/ddpg/tests/test_ddpg.py
Comment (kouroshHakha) [inline rllib/algorithms/ddpg/tests/test_ddpg.py:148]: same thing here. you have already specified num_steps_sampled_before_learning_starts in line 133

File: rllib/algorithms/dqn/dqn.py
Comment (kouroshHakha) [inline rllib/algorithms/dqn/dqn.py:387]: Is this code still needed? Can train_batch still be empty when you have waited for enough samples to get collected?

File: rllib/algorithms/dreamer/dreamer.py
Comment (kouroshHakha) [inline rllib/algorithms/dreamer/dreamer.py:291]: Can we remove this line?

File: rllib/algorithms/simple_q/simple_q.py
Comment (kouroshHakha) [inline rllib/algorithms/simple_q/simple_q.py:361]: same comment as above.

File: rllib/examples/offline_rl.py
Comment (kouroshHakha) [reply rllib/examples/offline_rl.py:41]: it shouldn't be broken anymore ?? see https://github.com/ray-project/ray/pull/27451

File: rllib/examples/two_trainer_workflow.py
Comment (kouroshHakha) [reply rllib/examples/two_trainer_workflow.py:115]: thanks for the clarification. I really like how you provide these extra PR explanations that answer most likely confusions preemptively. 👏

File: rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_replay_buffer.py
Comment (kouroshHakha) [inline rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_replay_buffer.py:176]: typo. Why is it not caught in CI? 🤔

File: rllib/utils/replay_buffers/utils.py
Comment (kouroshHakha) [inline rllib/utils/replay_buffers/utils.py:62]: Is this change of code related to this PR?

**Review by kouroshHakha — CHANGES_REQUESTED** [review]
Oh man!! this must have been a tough PR. Thanks a lot for this. There are a couple of major suggestions / todos before it can be merged. We can sync up over slack if you need more clarification. Also one more good PR practice that should become a habit: Please, Please let the reviewer resolve their own comments (If you mark them as resolved then the reviewer may miss their previous comment and it may indeed still be unresolved). It's also mandatory for the reviewer to also mark those comments that have been addressed as resolved and only leave those that are still open issues / not fully addressed. This way the communication is less prone to errors.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/alpha_zero/alpha_zero.py:119]: I've changed these comments.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/alpha_zero/alpha_zero.py:173]: The thing is that the on-policy algorithms don't use it and I can't think of a reason right now why they should ever need this parameter. What do you say?

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/alpha_zero/alpha_zero.py:226]: Changed.

File: rllib/algorithms/dreamer/dreamer.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/dreamer/dreamer.py:293]: Done!

File: rllib/algorithms/qmix/qmix.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/qmix/qmix.py]: Done. I also cleaned this from DQN and SimpleQ.

File: rllib/utils/replay_buffers/utils.py
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/utils.py:62]: This is from #27544 .
This PR surfaced a bug in RNNsac that we did not catch before.
I've separated the PRs but included #27544 in this PR for the tests to pass already.

File: release/rllib_tests/learning_tests/yaml_files/td3/td3-halfcheetahbulletenv-v0.yaml
Comment (kouroshHakha) [reply release/rllib_tests/learning_tests/yaml_files/td3/td3-halfcheetahbulletenv-v0.yaml:12]: **TODO**

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (kouroshHakha) [reply rllib/algorithms/alpha_zero/alpha_zero.py:173]: yeah but copying this code over and over also seems a bit of waste. We had replay buffer parameter before but it wasn't used by any of the on-policy algorithms. It makes more sense for it to be part of the base class.

File: rllib/algorithms/ddpg/tests/test_ddpg.py
Comment (kouroshHakha) [reply rllib/algorithms/ddpg/tests/test_ddpg.py:40]: **TODO**

File: rllib/algorithms/ddpg/tests/test_ddpg.py
Comment (kouroshHakha) [reply rllib/algorithms/ddpg/tests/test_ddpg.py:148]: **TODO**

File: rllib/utils/replay_buffers/utils.py
Comment (kouroshHakha) [reply rllib/utils/replay_buffers/utils.py:62]: I see. Do we have a unittest that would catch the old bug now?

File: rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_replay_buffer.py
Comment (kouroshHakha) [reply rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_replay_buffer.py:176]: **TODO**

**Comment by kouroshHakha** [comment]
@ArturNiederfahrenhorst This PR still needs to address all the **TODO**s

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/alpha_zero/alpha_zero.py:173]: I see, but this is the case for most of the parameters we use.
For example "lr_schedule" turns up in 6 different configs.
I strongly favor this staying here. Unneeded config options should explicitly not be part of any config. Even more so, a config object should error out upon initialization if user configures something that is not part of the algorithm in the long run.

File: rllib/algorithms/ddpg/tests/test_ddpg.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/ddpg/tests/test_ddpg.py]: Fixed. Also included L38 and L39.

File: rllib/algorithms/ddpg/tests/test_ddpg.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/ddpg/tests/test_ddpg.py:149]: Fixed!

File: rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_replay_buffer.py
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/tests/test_multi_agent_prioritized_replay_buffer.py:176]: This is because the MultiAgentReplayBuffers accept any kwargs for forward compatibility. If we want to make them swappable, they need to accept unneeded kwargs.

File: rllib/algorithms/alpha_zero/alpha_zero.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/alpha_zero/alpha_zero.py:173]: Absolutely, that's also the solution I thought of 👍

**Review by kouroshHakha — APPROVED** [review]
LGTM once test pass.

**Pull request merged**
